### PR TITLE
Add a note explaining how DATABASE_URL gets applied to current Rails environment [ci skip]

### PR DESCRIPTION
**Pull request by prathamesh-sonpatki**
### Summary

- This comment is an attempt to solve clarify how DATABASE_URL does
  not work for all environments by default and depends on Rails
  environment.

r? @schneems

### CODE REVIEW

**Comment by prathamesh-sonpatki** [comment]
Based on discussion in https://github.com/rails/rails/issues/24294

File: guides/source/configuring.md [hunk 1]
```diff
@@ -829,6 +829,8 @@ production:
 
 Now the behavior is clear, that we are only using the connection information in `ENV['DATABASE_URL']`.
 
+NOTE: When we are using `ENV['DATABASE_URL']` for connection management, it gets applied to the current Rails environment. This is in contrast with the `database.yml` where we have to explicitly add database configuration for different environments. By default, the `ENV['DATABASE_URL']` gets applied to `development` environment if Rails environment is not set.
```
Comment (nateberkopec) [inline guides/source/configuring.md]: > By default, the `ENV['DATABASE_URL']` gets applied to `development` environment if Rails environment is not set.

What are you trying to say here? I don't understand this.

**Comment by rafaelfranca** [comment]
I really hate this behavior. Since we have support to `url` in databasem.yml can we remove this special handling of DATABASE_URL once for all?

cc @schneems

**Comment by schneems** [comment]
I'm 👍  as long as there is a deprecation cycle and the default generated `production` for datbase.yml can use `DATABASE_URL` by default:

```
production:
  url: <%= ENV['DATABASE_URL'] %>
```

I mention this explicitly because it was set to that at one point, and then it got changed. If we take out the "magic" behavior and in a minor revision or two the default generated code gets changed i'll be pretty 😭 

This default config will work out of the box for anyone using it in production now. We talked with @matthewd at RubyConf about this and I think he is on board with the default config, deprecation and eventual removal of "magic" behavior.

File: guides/source/configuring.md (hunk 1 above)
Comment (nateberkopec) [inline guides/source/configuring.md]: Yeah, it may not be the fault of your explanation, it may just be that's behavior is kind of crazy 😆

**Comment by matthewd** [comment]
I changed my mind after we went back and looked at the current config. :worried:

I now believe the current behaviour is the best we're going to get: it allows the database to be set by the environment in a newly generated app (@schneems's requirement), and it leaves the default generated configuration using host/database/user/etc, rather than encouraging `url` even for in-`database.yml` configuration (@jeremy's requirement last time this was revisited, which I agree with).

I feel that the change to make `DATABASE_URI` only override the current environment's configuration, and not all possible configurations, was sufficient to bring the magic under control... our only remaining issue is `development`/`test` confusion, which we should be able to address more directly.

(The fact a command started under the `development`/default environment ends up running under `test` is already magical; that's where further magic belongs. But we should indeed improve the experience, not paper over it with a doc change people won't read.)

**Comment by prathamesh-sonpatki** [comment]
@matthewd I had looked into setting test environment for test database rake tasks but I think the test database rake tasks traditionally don't depend on the Rails environment, they just pick up the test database configuration assuming that it is always present when they are run. 

I opened https://github.com/rails/rails/pull/27102 to raise helpful error in cases when somebody tries to run test database rake tasks without configuring test database. The error is applicable to both DATABASE_URL and database.yml.  This will give an error saying that 'you have not configured the test database`. Should that PR along with this documentation note be enough?

**Comment by prathamesh-sonpatki** [comment]
@matthewd What do you think about https://gist.github.com/prathamesh-sonpatki/7aaf542e26d800d15a93d6d7ff7c7df1 

After that the test tasks will be executed in "test" environment.

**Comment by prathamesh-sonpatki** [comment]
We don't even need #27102 after that.

**Comment by matthewd** [comment]
Hmm.. I think I prefer #27102. I guess @schneems's DB+env check would protect us anyway, but it still seems better to error, on the assumption that the set `DATABASE_URL` probably belongs to development, and not test.

**Comment by prathamesh-sonpatki** [comment]
@matthewd Sounds good! Thoughts about executing the test rake tasks in test environment though? Would it be good thing to have apart from this DATABASE_URL mismatch?

**Comment by jrochkind** [comment]
Is this added doc still accurate? If so, would it make sense to merge it? Or decide not to and close the ticket?

There is discussion above sparked by the doc request to change behavior; not sure if that ever went anywhere, but it's really a different discussion then adding doc. If the doc is still accurate, I think maybe it should just be merged?

I agree DATABASE_URL env handling is under-doc'd, and could use some.

File: guides/source/configuring.md (hunk 1 above)
Comment (jrochkind) [reply guides/source/configuring.md]: I think the last sentance can just be eliminated, and the previous sentences explain it well enough. I also thought the last sentance was appearing to contradict the previous sentances, although now I see what you mean.

**Pull request closed without merging**